src/lib/containers/entity_finder/EntityFinder.tsx

```diff
@@ -0,0 +1,320 @@
+          size={'sm'}
+          icon={faTimes}
+        />
+      </span>
+      <span>{pathString ? pathString + '/' : ''}</span>
+      <span style={{ fontWeight: 'bold' }}>{entityName}</span>
+      {entity.targetVersionNumber && (
+        <span> (Version {entity.targetVersionNumber})</span>
+      )}
+    </>
+  )
+}
+
+type EntityFinderProps = {
+  sessionToken: string
+  initialContainerId: string // The SynID of the entity that should open by default. This dictates the 'Current Project'
+  selectMultiple: boolean
+  onSelectedChange: (selected: Reference[]) => void
+  showVersionSelection?: boolean
+  showTypes: EntityType[]
+  selectableTypes?: EntityType[]
+}
+export const EntityFinder: React.FunctionComponent<EntityFinderProps> = ({
+  sessionToken,
+  initialContainerId,
+  selectMultiple,
+  onSelectedChange,
+  showVersionSelection = true,
+  showTypes,
+  selectableTypes = Object.values(EntityType),
+}) => {
+  const [selectedEntities, setSelectedEntities] = useState<Reference[]>([])
+
+  const [searchActive, setSearchActive] = useState(false)
+  const [searchTerms, setSearchTerms] = useState<string[]>()
+  const [searchByIdResults, setSearchByIdResults] = useState<EntityHeader[]>([])
+  const [
+    configFromTreeView,
+    setConfigFromTreeView,
+  ] = useState<EntityDetailsListDataConfiguration>()
+
+  const isSelected = (entity: Reference) => {
+    return selectedEntities.some(
+      s =>
+        s.targetId === entity.targetId &&
+        s.targetVersionNumber === entity.targetVersionNumber,
+    )
+  }
+
+  const otherVersionSelected = (entity: Reference) => {
+    return selectedEntities.some(
+      s =>
+        s.targetId === entity.targetId &&
+        s.targetVersionNumber !== entity.targetVersionNumber,
+    )
+  }
+
+  const toggleSelection = (entity: Reference) => {
+    if (isSelected(entity)) {
+      // remove from selection
+      setSelectedEntities(
+        selectedEntities.filter(e => e.targetId !== entity.targetId),
+      )
+    } else if (otherVersionSelected(entity)) {
+      // replace with selected version
+      setSelectedEntities([
+        ...selectedEntities.filter(e => e.targetId !== entity.targetId),
+        entity,
+      ])
+    } else {
+      // add to selection
+      if (!selectMultiple) {
+        setSelectedEntities([entity])
+      } else {
+        setSelectedEntities([
+          ...selectedEntities.filter(s => s.targetId !== entity.targetId),
+          entity,
+        ])
+      }
+    }
+  }
+
+  useEffect(() => {
+    onSelectedChange(selectedEntities)
+  }, [onSelectedChange, selectedEntities])
+
+  useEffect(() => {
+    if (searchTerms?.length === 1) {
+      const synIdMatch = searchTerms[0].match(SYNAPSE_ENTITY_ID_REGEX)
+      if (synIdMatch) {
+        SynapseClient.getEntityHeaders(
+          [
+            {
+              targetId: synIdMatch[1],
+              targetVersionNumber: synIdMatch[2]
+                ? parseInt(synIdMatch[2])
```

Comment:
lot of potential for unhandled errors in the code, but I may as well point one out (parsing error).